common/aws/s3/client.go

```diff
@@ -159,3 +175,162 @@ func (s *client) ListObjects(ctx context.Context, bucket string, prefix string)
 	}
 	return objects, nil
 }
+
+func (s *client) CreateBucket(ctx context.Context, bucket string) error {
+	_, err := s.s3Client.CreateBucket(ctx, &s3.CreateBucketInput{
+		Bucket: aws.String(bucket),
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (s *client) FragmentedUploadObject(
+	ctx context.Context,
+	bucket string,
+	key string,
+	data []byte,
+	fragmentSize int) error {
+
+	fragments, err := BreakIntoFragments(key, data, s.cfg.FragmentPrefixChars, fragmentSize)
+	if err != nil {
+		return err
+	}
+	resultChannel := make(chan error, len(fragments))
+
+	ctx, cancel := context.WithTimeout(ctx, s.cfg.FragmentWriteTimeout)
+	defer cancel()
+
+	for _, fragment := range fragments {
+		fragmentCapture := fragment
+		s.pool.Submit(func() {
+			s.fragmentedWriteTask(ctx, resultChannel, fragmentCapture, bucket)
+		})
```

Comment:
Any benefits of having our own worker pool implementation instead of using https://pkg.go.dev/golang.org/x/sync/errgroup?

Reply:
errgroup has a limit feature now! https://pkg.go.dev/golang.org/x/sync/errgroup#Group.SetLimit

Reply:
Neat. Switched over to using errgroup.